### ISSUE TITLE
image does not work with actions/checkout

### ISSUE THREAD

**Issue by JasonGross** [coq-community/docker-coq#60]
When running `actions/checking` with `container: coqorg/coq:dev`, [I see](https://github.com/mit-plv/fiat-crypto/actions/runs/7320568634/job/19939855942#step:7:18)
```
/usr/bin/docker exec  d2f4cc80928d9913b66a607fd95b3abc47ed3b5a78234ed06151b3bcd6260342 sh -c "cat /etc/*release | grep ^ID"
node:internal/fs/sync:78
  return binding.openSync(
                 ^

Error: EACCES: permission denied, open '/__w/_temp/_runner_file_commands/save_state_e59de318-2ebd-4430-adee-f74fc8e00eb1'
    at Object.open (node:internal/fs/sync:78:18)
    at Object.openSync (node:fs:565:17)
    at Object.writeFileSync (node:fs:2288:35)
    at Object.appendFileSync (node:fs:2350:6)
    at Object.issueFileCommand (/__w/_actions/actions/checkout/v4/dist/index.js:2967:8)
    at Object.saveState (/__w/_actions/actions/checkout/v4/dist/index.js:2884:31)
    at 8647 (/__w/_actions/actions/checkout/v4/dist/index.js:2343:10)
    at __nccwpck_require__ (/__w/_actions/actions/checkout/v4/dist/index.js:18273:43)
    at 2565 (/__w/_actions/actions/checkout/v4/dist/index.js:146:34)
    at __nccwpck_require__ (/__w/_actions/actions/checkout/v4/dist/index.js:18273:43) {
  errno: -13,
  code: 'EACCES',
  syscall: 'open',
  path: '/__w/_temp/_runner_file_commands/save_state_e59de318-2ebd-4430-adee-f74fc8e00eb1'
}

Node.js v20.8.1
```

How do I get this image working?

**Comment by JasonGross** [comment]
Would it be feasible to change the uid and gid to 1001 and 127 respectively to match github runners?  Or at least to make `docker run -it --user root coqorg/coq:dev /bin/bash` work (rather than fail with ```[ERROR] Opam has not been initialised, please run `opam init'```) by symlinking `/home/root/.opam` and `/home/root/.bashrc` to the corresponding files in `/home/coq`?

**Comment by JasonGross** [comment]
See also https://github.com/actions/checkout/issues/1576

**Comment by erikmd** [comment]
Thanks @JasonGross 

> Would it be feasible to change the uid and gid to 1001 and 127 respectively to match github runners? Or at least to make `docker run -it --user root coqorg/coq:dev /bin/bash` work (rather than fail with `` [ERROR] Opam has not been initialised, please run `opam init' ``) by symlinking `/home/root/.opam` and `/home/root/.bashrc` to the corresponding files in `/home/coq`?

BTW where is documented this GID=127 ? [IIRC,](https://github.com/coq-community/docker-coq-action/#permissions) the github action workdir has GID=116

From the time being, I partly replied here:

https://github.com/actions/checkout/issues/1576#issuecomment-1872360027

**Comment by JasonGross** [comment]
> Not exactly, there is [an entrypoint](https://github.com/coq-community/docker-base/blob/7e36b663d66c4529e3cd7022ede58af2e884719c/base/single/Dockerfile#L100), but you can certainly override it:
> 
> `docker run -it --user root coqorg/coq:dev --entrypoint=/bin/sh`

I see
```
$ docker run -it --user root coqorg/coq:dev --entrypoint=/bin/sh
[WARNING] Running as root is not recommended
[ERROR] Opam has not been initialised, please run `opam init'
```
Does this invocation actually work for you?

**Comment by JasonGross** [comment]
> * what are the main use cases you aim at doing this? (maybe among these new use cases, some could be added to docker-coq-action)

I'd like to have a single docker container running in the background that is shared between multiple steps of the same job, much like having `container: coqorg/coq:dev` or similar for the entire job would do.

Also, I would like to be able to share more code between testing on Docker coq, Alpine, Arch Linux, Debian, etc.  If I can set it up so that the only difference between these platforms is
1. passing a different argument to `container:` in my workflow, and
2. invoking a different script to set up the system dependencies, etc, on the container

that would be ideal.

**Comment by JasonGross** [comment]
> * do you know where is documented the "init script" of `actions/checkout` that requires `USER root` ?

There's a "Detailed Summary" section of https://github.com/actions/checkout/issues/956 that explains why having either `USER root` or uid 1001 results in success.  Basically, the issue is that some parts of the checkout action run outside the docker container, while other parts run inside, and the parts that run inside avoid "permission denied" only if either uid matches the outside uid or the user is root.  I personally think actions/checkout should fix this on their side and support the use of sudo to escalate privileges in the container.

**Comment by erikmd** [comment]
> Does this invocation actually work for you?

Yes, but with the `--entrypoint` option earlier in the command line (sorry for the "typo" in my previous comment):

`docker run -it --user root --entrypoint /bin/bash coqorg/coq:latest`

**Comment by erikmd** [comment]
> I'd like to have a single docker container running in the background that is shared between multiple steps of the same job, much like having container: coqorg/coq:dev or similar for the entire job would do.

OK, good point.

> Also, I would like to be able to share more code between testing on Docker coq, Alpine, Arch Linux, Debian, etc. If I can set it up so that the only difference between these platforms is (…) invoking a different script to set up the system dependencies, etc, on the container

I'm not sure I follow you: the Docker-Coq containers do have a Linux distro by themselves (Debian)

**Comment by JasonGross** [comment]
Hmmm, I'm not sure how to get the entry point working.  According to [stack overflow](https://stackoverflow.com/a/59903328/377022), it doesn't seem to be possible to override the entrypoint when specifying a container to use.  Maybe I need to create a custom docker image here or something?  



> > Also, I would like to be able to share more code between testing on Docker coq, Alpine, Arch Linux, Debian, etc. If I can set it up so that the only difference between these platforms is (…) invoking a different script to set up the system dependencies, etc, on the container
> 
> I'm not sure I understand what you mean! the Docker-Coq containers do have a Linux distro by themselves (Debian)

I mean that in addition to testing my code against coqorg/coq:dev, I also want to test my code against the system-packaged version of Coq on Debian, Alpine, ArchLinux, etc.  It'd be nice to reuse workflow code between these platforms

**Comment by erikmd** [comment]
Hi!

> it doesn't seem to be possible to override the entrypoint when specifying a container to use. Maybe I need to create a custom docker image here or something?

I don't think so (anyway, that would really looks like some overengineering to me… IMO, a default entrypoint such as this one should never be a blocker, nor justify creating and maintaining custom images just for removing the entrypoint)

can you test this?
```
docker run -it --user root -e OPAMROOT=/home/coq/.opam coqorg/coq:dev /bin/bash -c 'eval $(opam env); exec /bin/bash' bash
```

**Comment by erikmd** [comment]
> I mean that in addition to testing my code against coqorg/coq:dev, I also want to test my code against the system-packaged version of Coq on Debian, Alpine, ArchLinux, etc. It'd be nice to reuse workflow code between these platforms

Ah OK, I understand better.

But IINM, regarding workflow code reuse, I guess you already have a bash script, compatible with both scenarios…
so isn't the [`custom_script:`](https://github.com/coq-community/docker-coq-action/#custom_script) feature of **`docker-coq-action`** enough for your use case?
otherwise, what is missing/blocking?

**Comment by JasonGross** [comment]
> just for removing the entrypoint; one should either fix the limitation of the platform

Yeah, I think that would be ideal, if I can convince GH Actions to support this...


> can you test something like this?
> 
> ```
> docker run -it --user root -e OPAMROOT=/home/coq/.opam coqorg/coq:dev /bin/bash -c 'eval $(opam env); exec /bin/bash' bash
> ```

This works locally, at least, thanks!

> But IINM, regarding workflow code reuse, I guess you already have a bash script, compatible with both scenarios…
> so isn't the [`custom_script:`](https://github.com/coq-community/docker-coq-action/#custom_script) feature of **`docker-coq-action`** enough for your use case?
> otherwise, what is missing/blocking?

For workflow code reuse, yes.  The main missing feature is not having to reinstall dependencies / set up the container at every job step, and still getting nice top-level grouping.

**Comment by erikmd** [comment]
AFAICT, currently, the image (not the container) is already reused: thanks to these lines:

https://github.com/coq-community/docker-coq-action/blob/1c523fff4617a5422a324720a2fdcf4a96ef3e6a/entrypoint.sh#L198-L204

the pull of the same `$COQ_IMAGE` is only done once.

**Comment by erikmd** [comment]
Hi @JasonGross, Thanks for your feedback.

I believe it would be feasible to add some YAML Boolean in docker-coq-action's interface, e.g.:

```yaml
steps:
  - uses: actions/checkout@v3
  - uses: coq-community/docker-coq-action@v1
    with:
      opam_file: 'folder/coq-proj.opam'
      coq_version: ${{ matrix.coq_version }}
      ocaml_version: ${{ matrix.ocaml_version }}
      non_ephemeral: true
```

Concretely, this would remove the `docker run --rm` CLI option and pass the `-d` option instead.

Questions (Cc @Zimmi48, feel free to react as well if ever you have comments):

1. does the `non_ephemeral` keyword looks good to you? or do you have other suggestions? (AFAIAC, I prefer keywords that suggest that this option is not the default recommended one, given for example that ephemeral containers are generally recommended in CI)
2. don't you think it would be a good invariant to require that two successive `coq-community/docker-coq-action@v1` steps with `non_ephemeral: true` have the same Docker image name ?, e.g. so that the following raise an error:
   ```
   - uses: coq-community/docker-coq-action@v1
     with:
       non_ephemeral: true
       custom_image: 'coqorg/coq:dev'
       custom_script: |
         …
   - uses: coq-community/docker-coq-action@v1
     with:
       non_ephemeral: true
       custom_image: 'coqorg/coq:latest'
       custom_script: |
         …
     ```
   after some quick test, this invariant could be checked using `docker inspect -f '{{.Config.Image}}' COQ`
3. Currently, the container name is fixed: "COQ" (implying only one container can run in the background);
   would you be interested to run several different containers in the background?
4. If the answer to 3. is yes, we could restrict the "COQ" container name to ephemeral containers,
   but we'd need to specify the container naming convention:  
   (4.1) either deduced from the image, replacing `/` with `--` and `:` with `..` for instance (but this is ugly),  
   (4.2) or specified by the user with a `container_name: '…'` keyword (with '…' ≠ 'COQ')
5. Finally, a post-processing step would be required to trigger `docker rm -f` at the end of the job.

Thoughts? other questions?

**Comment by JasonGross** [comment]
I like the direction this goes in, but I think the configuration could be simplified to just having a single `container_name` option.
- The container is ephemeral iff the `container_name` is `''`, in which case the container is named `COQ`.
- Error on `container_name: COQ`.
- If `container_name` is given explicitly and the container already exists, maybe error if the image/version is explicitly given but does not match the existing container?

**Comment by erikmd** [comment]
Hi @JasonGross, thanks for your update! I agree with your suggestion.
Maybe it's unneeded to have a keyword that verbose such as `non_ephemeral_container_name`,
so let's stick to `container_name` for now.
I will prepare a PR in the upcoming days (not 100% sure that the feature is feasible but I'm confident enough).
Then, @Zimmi48 will be able to review it (especially for the accompanying documentation and the `container_name` phrasing) before we obtain a new release.

**Comment by erikmd** [comment]
Hi @JasonGross, given this issue led to a docker-coq-action wishlist, I'm migrating the issue now.
